python/lsst/ctrl/mpexec/cli/script/purge.py

```diff
@@ -0,0 +1,261 @@
+# This file is part of ctrl_mpexec.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+import itertools
+from typing import List, Union
+
+from lsst.daf.butler import Butler, CollectionType
+from lsst.daf.butler.registry import MissingCollectionError
+
+from .confirmable import ConfirmableResult
+
+advice = (
+    'Use "butler collection-chain --mode remove" to remove this collection from its parent or\n'
+    'use "butler remove-collections" to remove that parent entirely.'
+)
+
+
+class ChildHasMultipleParentsFailure:
+    def __init__(self, child: str, parents: List[str]):
+        self.child = child
+        self.parents = parents
+
+    def __str__(self):
+        parents = ", ".join([f'"{p}"' for p in self.parents])
+        return f'Collection "{self.child}" is in multiple chained collections: {parents}.\n' + advice
+
+
+class TopCollectionHasParentsFailure:
+    def __init__(self, collection: str, parents: List[str]):
+        self.collection = collection
+        self.parents = parents
+
+    def __str__(self):
+        parents = ", ".join([f'"{p}"' for p in self.parents])
+        return (
+            f'The passed-in collection "{self.collection}" must not be contained in other collections but '
+            f"is contained in collection(s) {parents}.\n" + advice
+        )
+
+
+class TopCollectionIsNotChianedFailure:
+    def __init__(self, collection: str, collection_type: CollectionType):
+        self.collection = collection
+        self.collection_type = collection_type
+
+    def __str__(self):
+        return (
+            "The passed-in collection must be a CHAINED collection; "
+            f'"{self.collection}" is a {self.collection_type.name} collection.'
+        )
+
+
+class TopCollectionNotFoundFailure:
+    def __init__(self, collection: str):
+        self.collection = collection
+
+    def __str__(self):
+        return f'The passed-in colleciton "{self.collection}" was not found.'
+
+
+class PurgeResult(ConfirmableResult):
+    def __init__(self, butler_config: str):
+        self.runs_to_remove = []
+        self.chains_to_remove = []
+        self.others_to_remove = []
```

Comment:
Making these `set`s instead of `list`s seems like would reflect the usage better, unless you're concerned about printing them for confirmation in a way that maintains order.